switchbot/devices/curtain.py

```diff
@@ -54,6 +57,18 @@ def __init__(self, *args: Any, **kwargs: Any) -> None:
         self._settings: dict[str, Any] = {}
         self.ext_info_sum: dict[str, Any] = {}
         self.ext_info_adv: dict[str, Any] = {}
+        self._is_opening: bool = False
+        self._is_closing: bool = False
+
+    def _set_parsed_data(
+        self, advertisement: SwitchBotAdvertisement, data: dict[str, Any]
+    ) -> None:
+        """Set data."""
+        _new_data = replace(
+            advertisement, data=self._sb_adv_data.data | {"data": data}
+        )
+        self._update_motion_direction(_new_data.data["data"].get("inMotion"), self._get_adv_value("position"), _new_data.data["data"].get("position"))
```

Comment:
```suggestion
        
        _new_data = replace(
            advertisement, data=self._sb_adv_data.data | {"data": data}
        )
        self._update_motion_direction(_new_data.data["data"].get("inMotion"), self._get_adv_value("position"), _new_data.data["data"].get("position"))
```
```suggestion
        in_motion = data["inMotion"]
        previous_position = self._get_adv_value("position")
        new_position = data["position"]
        self._update_motion_direction(in_motion, previous_position, new_position)
```

Reply:
Ok I see what you're saying. My understanding was I need something like `self._sb_adv_data.data | {"data": data}` because reading through the code sometimes we get partial advertisements? Meaning it's possible the position value is missing entirely from the most recent advertisement and the union makes it so we get the existing value from sb_adv_data. E.g. maybe the current advertisement is wholesale missing the inMotion value, so we assume, in its absence, the value is the same from the last advertisement. I'm not finding the comment where I could swear I read this so maybe I hallucinated it. You know the library much better than I do. Let me know if what I said makes sense and if the change I mentioned in https://github.com/Danielhiversen/pySwitchbot/pull/226#discussion_r1446831874 makes sense otherwise I'll accept your suggestion.

Reply:
That seems like it would be a bug in the firmware to send invalid advertisements.  I'm pretty sure it wouldn't get to this point as the parser would reject it if its was short bytes